### ISSUE TITLE
Explain the discrepancy in metric values between the old SCT Course slides and current processing

### ISSUE THREAD

**Issue by joshuacwnewton** [spinalcordtoolbox/spinalcordtoolbox#4297]
This is less of an issue and more of a question:

- Previously, the SCT course was structured as follows:
   1. SC Segmentation + Labeling 
   2. Registration to template + warping
   3. Compute shape metrics on segmentation using `-vertfile PAM50_levels.nii.gz`
- Now, the SCT course is structured as follows:
   1. SC Segmentation + Labeling
   2. Compute shape metrics on segmentation using `-vertfile t2_seg_labeled.nii.gz`
   3. Registration to template + warping

The change in results is not insignificant. For example, computing the mean CSA across C3:C4 vertebral levels changes from 78.0137 -> 74.1357. 

This change could maybe be explained by a difference in the labeled vertebral slices (e.g. ~2 fewer slices on the top and ~2 more slices on the bottom, causing the average to shift towards the thinner region of the cord). But, I would like to actually properly explain this discrepancy, just so that we are not just updating numbers blindly.

**Comment by joshuacwnewton** [comment]
> This change could easily be explained by a difference in the labeled vertebral slices (e.g. ~2 fewer slices on the top and ~2 more slices on the bottom, causing the average to shift towards the thinner region of the cord).

This is _not_ the case! Instead, the CSA values are much smaller regardless of which slices are specified. 

This is a bit confusing to me. Aren't we using the same `t2_seg.nii.gz` image in both cases? Why would it matter which `-vertfile` is used? 

Unless... is `-vertfile` being used for the CSA computation? In that case, why are we even specifying `t2_seg.nii.gz`... :thinking:

**Comment by joshuacwnewton** [comment]
This may also be due to a change in the actual data we used between versions of the course! Hmm...

**Comment by jcohenadad** [comment]
Thank you for digging into this @joshuacwnewton ! Indeed 4mm^2 is definitely not insignificant and could not (I think) be explained by a different of 1-2 slices. So, as you said, the change could be due to:
- change in the data (but has it changed since the last course? I don't think so)
- change in SCT (hum, that would be more problematic)
- change in the segmentation method used (if we played around with propseg and deepseg, and now we got rid of propseg, maybe that explains it? IIRC we didn't do deepseg on the T2 previously)

**Comment by joshuacwnewton** [comment]
Very good points! I will make sure to explore each possibility thoroughly. :)

**Comment by joshuacwnewton** [comment]
I've updated the title of this issue because I've realized that A) the discrepancies go beyond CSA and B) the discrepancies seem to be platform-dependent (see [this comment thread](https://docs.google.com/presentation/d/1t40Fd0fS0SwWR5FU_GWKEvHkB9d96LVddLQW6L3QAx0/edit?disco=AAAA_ihZnog)).

**Comment by joshuacwnewton** [comment]
Small note: I've been working on this now as a post-SCT Course task. I'll make an update comment when I'm finished, but it's taking a bit longer than I'd hoped, since it involves comparing across SCT Course versions + SCT versions + Operating Systems (Windows/Ubuntu/etc.)

But, if it's quiet from me, then you know what I'm working on, hehe.

**Comment by joshuacwnewton** [comment]
## T2 CSA

> * change in the segmentation method used (if we played around with propseg and deepseg, and now we got rid of propseg, maybe that explains it? IIRC we didn't do deepseg on the T2 previously)

This is precisely the explanation for the large discrepancy reported above. If I continue with the usual course steps (running `./batch_single_subject.sh` from each of the respective course datasets), but modifying only the initial T2 seg command to test both `{sct_propseg, sct_deepseg_sc}`, then the results are as follows:

| Source                                   | sct_propseg      | sct_deepseg_sc    | 
| ---------------------------------------- | ---------------- | ----------------- |
| 2021 Keynote                             | 78.0137361494396 |                   |
| data `r20211116`, SCT v5.4, Ubuntu 22.04 | 78.0933348005592 | 74.28722611282844 |
| data `r20211116`, master, Ubuntu 22.04   | 78.3304486395692 | 74.31003297750931 |
| data `r20231120`, SCT v5.4, Ubuntu 22.04 | 78.1198714916236 | 74.13571787652657 |
| data `r20231120`, master, Ubuntu 22.04   | 78.2609567327718 | 74.15906791196547 |

And, for the minor discrepancies shown above, I think this can be chalked up to A) version differences and B) switching from warped PAM50 level files to sct_label_vertebrae level file.

## MTR in WM

Here, the discrepancies exist between time periods and versions, but also between OSs:

### SCT Course 2021 vs 2023

It looks like the values have changed both over versions of SCT and versions of the SCT course (sample value: slice 1 of mtr.nii.gz in wm):

| Source                                   | Size[vox]         | MAP()             | 
| ---------------------------------------- | ----------------- | ----------------- |
| 2021 Keynote                             | 48.1509           | 57.2138           |
| data `r20211116`, SCT v5.4, Ubuntu 22.04 | 45.28498611873558 | 57.25200977669559 |
| data `r20211116`, master, Ubuntu 22.04   | 44.77387889726627 | 57.16024181366148 |
| data `r20231120`, SCT v5.4, Ubuntu 22.04 | 45.91386298871404 | 56.64487691057087 |
| data `r20231120`, master, Ubuntu 22.04   | 46.03516842444189 | 56.92654643566894 |

It seems like both the version of SCT and the version of `sct_tutorial_data` have an effect on the results here. But, experientially, MTR in WM has been one of the more sensitive metrics that we test in our pipelines, and is one we often have to change in `batch_processing.sh`. So, I'm not surprised to see a lot of variability over the last 2 years.

Truthfully I'm not really sure what actionable steps I can take by comparing 2021 to 2023? There's no "smoking gun" like there was with the CSA.

### SCT Course 2023: Ubuntu vs. Windows

| Source                                   | Size[vox]         | MAP()             | 
| ---------------------------------------- | ----------------- | ----------------- |
| data `r20231120`, master, Ubuntu 22.04   | 46.03516842444189 | 56.92654643566894 |
| data `r20231120`, 6.1, Windows           | 45.68361106818536 | 56.88810748094038 |
| data `r20231120`, Mathieu's comment      | 47.29891978527233 | 56.91225824981293 |

This discrepancy, on the other hand, is something we can do about in the here and now. 

I was under the impression that there [were no cross-platform discrepancies between Ubuntu and Windows](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3194#issuecomment-1099851235). So, if there _are_ OS-specific differences between Ubuntu and Windows, it means that our `batch_processing.sh` is missing a key test case and thus can be improved.

Given that we already have several issues pertaining to OS-specific reproducibility with batch_processing.sh, I think it might be better to investigate these discrepancies in their respective issues, though:

- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3250
- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3194
- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3508

In other words, the action items would be:

- Work towards harmonizing `batch_processing.sh` and `batch_single_subject.sh` such that the `batch_processing.sh` test is actively checking the metrics displayed in the SCT course slides.
   - Either by adding missing checks to batch_processing.sh, or
   - By updating existing commands if there are differences between the two scripts
- Check to see if any of the newly-added metrics vary across OSs (and if they do, debug why)

This way, we can see the changes in the slides coming ahead of time, way before we actually need to update the course slides.

**Comment by jcohenadad** [comment]
amazing and important detective work @joshuacwnewton ❤️
